tools/src/wyvern/target/corewyvernIL/decl/DefDeclaration.java

```diff
@@ -79,6 +88,16 @@ public DeclType typeCheck(TypeContext ctx, TypeContext thisCtx) {
 		for (FormalArg arg : formalArgs) {
 			methodCtx = methodCtx.extend(arg.getName(), arg.getType());
 		}
+
+		for (String freeVar : this.getFreeVariables()) {
+			ValueType t = (new Variable(freeVar)).typeCheck(methodCtx);
+			if (t instanceof StructuralType) {
```

Comment:
Avoid instanceof by making isResource() a member of ValueType (and providing implementations for other kinds of ValueTypes).